xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,239 @@
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @param <E>
+ * @since 10.0RC1
+ */
+public class XWikiAttachmentList extends AbstractNotifyOnUpdateList<XWikiAttachment>
+{
+
+    private final Map<String, XWikiAttachment> map = new ConcurrentSkipListMap<String, XWikiAttachment>();
+
+    private final XWikiDocument document;
+
+    /**
+     * Initializes the map.
+     * 
+     * @since 10.0RC1
+     */
+    public XWikiAttachmentList(XWikiDocument document)
+    {
+        super(new ArrayList<XWikiAttachment>());
+        this.document = document;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        this.list = new ArrayList<XWikiAttachment>(map.values());
+        onUpdate();
+        added(attachment);
+        return true;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index, XWikiAttachment attachment)
+    {
+        add(attachment);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public void clear()
+    {
+        super.clear();
+        map.clear();
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        this.list = new ArrayList<XWikiAttachment>(map.values());
+        onUpdate();
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends XWikiAttachment> c)
+    {
+        return addAll(c);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.list.get(index).getFilename());
+        if (removedAttachment != null) {
+            removedAttachment = super.remove(index);
+            document.setMetaDataDirty(true);
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return XWikiAttachment that was removed or null if not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        int index = list.indexOf(attachment);
```

Comment:
It would be faster to use the Map for this: call Map#put() and check if the return is the same than the passed XWikiAttachment and if they are different update the list, notify the XWikiDocument, return true, etc.